lib/posix/eventfd.c

```diff
@@ -246,51 +262,181 @@ static const struct fd_op_vtable eventfd_fd_vtable = {
 	.ioctl = eventfd_ioctl_op,
 };
 
-int eventfd(unsigned int initval, int flags)
+/* common to both eventfd_read_op() and eventfd_write_op() */
+static ssize_t eventfd_rw_op(void *obj, void *buf, size_t sz,
+			     int (*op)(struct eventfd *efd, eventfd_t *value))
 {
-	struct eventfd *efd = NULL;
-	int fd = -1;
-	int i;
+	int err;
+	ssize_t ret;
+	k_spinlock_key_t key;
+	struct eventfd *efd = obj;
+	struct k_mutex *lock = NULL;
+	struct k_condvar *cond = NULL;
 
-	if (flags & ~EFD_FLAGS_SET) {
+	if (sz < sizeof(eventfd_t)) {
 		errno = EINVAL;
 		return -1;
 	}
 
-	k_mutex_lock(&eventfd_mtx, K_FOREVER);
+	if (buf == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	key = k_spin_lock(&efd->lock);
+
+	if (!eventfd_is_blocking(efd)) {
+		/*
+		 * Handle the non-blocking case entirely within this scope
+		 */
+		ret = op(efd, buf);
+		if (ret < 0) {
+			errno = -ret;
+			ret = -1;
+		} else {
+			ret = sizeof(eventfd_t);
+		}
+
+		goto unlock_spin;
+	}
+
+	/*
+	 * Handle the blocking case below
+	 */
+	__ASSERT_NO_MSG(eventfd_is_blocking(efd));
+
+	if (k_is_in_isr()) {
+		/* not covered by the man page, but necessary in Zephyr */
+		errno = EWOULDBLOCK;
+		ret = -1;
+		goto unlock_spin;
+	}
+
+	err = (int)z_get_obj_lock_and_cond(obj, &eventfd_fd_vtable, &lock, &cond);
+	__ASSERT((bool)err, "z_get_obj_lock_and_cond() failed");
+	__ASSERT_NO_MSG(lock != NULL);
+	__ASSERT_NO_MSG(cond != NULL);
+
+	/* do not hold a spinlock when taking a mutex */
+	k_spin_unlock(&efd->lock, key);
+	err = k_mutex_lock(lock, K_FOREVER);
+	__ASSERT(err == 0, "k_mutex_lock() failed: %d", err);
 
-	for (i = 0; i < ARRAY_SIZE(efds); ++i) {
-		if (!(efds[i].flags & EFD_IN_USE)) {
-			efd = &efds[i];
+	while (true) {
+		/* retake the spinlock */
+		key = k_spin_lock(&efd->lock);
+
+		ret = op(efd, buf);
```

Comment:
How safe is it to allow a caller defined function to be called while holding the spin lock?

I guess it is somewhat moot since it is clearly necessary to lock here. Just asking for curiosity.

Reply:
This function is static and the only functions that are being passed in are `eventfd_read_locked()` and `eventfd_wrote_locked()`